pkg/storage/utils/decomposedfs/upload/upload.go

```diff
@@ -0,0 +1,446 @@
+		// etag still matches before finishing the upload.
+		if ifMatch, ok := upload.Info.MetaData["if-match"]; ok {
+			var targetEtag string
+			targetEtag, err = node.CalculateEtag(n.ID, fi.ModTime())
+			if err != nil {
+				return errtypes.InternalError(err.Error())
+			}
+			if ifMatch != targetEtag {
+				return errtypes.Aborted("etag mismatch")
+			}
+		}
+
+		// FIXME move versioning to blobs ... no need to copy all the metadata! well ... it does if we want to version metadata...
+		// versions are stored alongside the actual file, so a rename can be efficient and does not cross storage / partition boundaries
+		versionsPath = upload.lu.InternalPath(spaceID, n.ID+node.RevisionIDDelimiter+fi.ModTime().UTC().Format(time.RFC3339Nano))
+
+		// we unset the processing flag before moving
+		// NOTE: this leads to a minimal time the node has no processing flag
+		if err := n.UnmarkProcessing(); err != nil {
+			sublog.Error().Err(err).Msg("Decomposedfs: could not unmark processing")
+			return err
+		}
+		// This move drops all metadata!!! We copy it below with CopyMetadata
+		// FIXME the node must remain the same. otherwise we might restore share metadata
+		if err = os.Rename(targetPath, versionsPath); err != nil {
+			sublog.Err(err).
+				Str("binPath", upload.binPath).
+				Str("versionsPath", versionsPath).
+				Msg("Decomposedfs: could not create version")
+			return err
+		}
+
+		// NOTE: In case there is an existing version we have no processing flag on the targetPath,
+		// as we just moved that file. We need to create an empty file again
+		// TODO: that means that there is a short amount of time when there is no targetPath or no processing flag
+		// If clients query in exactly that moment the file will be gone from their PROPFIND
+		// How can we omit this issue? How critical is it?
+		if _, err := os.Create(targetPath); err != nil {
+			sublog.Error().Err(err).Msg("Decomposedfs: could not create file")
+			return err
+		}
+		// and set the processing flag on this
+		if err := n.MarkProcessing(); err != nil {
+			sublog.Error().Err(err).Msg("Decomposedfs: could not mark processing")
+			return err
+		}
+	}
+
+	// upload the data to the blobstore
+	file, err := os.Open(upload.binPath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	err = upload.tp.WriteBlob(n, file)
+	if err != nil {
+		return errors.Wrap(err, "failed to upload file to blostore")
+	}
+
+	// now truncate the upload (the payload stays in the blobstore) and move it to the target path
+	// TODO put uploads on the same underlying storage as the destination dir?
+	if err = os.Truncate(upload.binPath, 0); err != nil {
+		sublog.Err(err).
+			Msg("Decomposedfs: could not truncate")
+		return err
+	}
+	if err = os.Rename(upload.binPath, targetPath); err != nil {
+		sublog.Error().Err(err).Msg("Decomposedfs: could not rename")
+		return err
+	}
+	// the rename dropped the "processing" status - we need to set it again
```

Comment:
hm, we could copy the metadata from the node to the upload.binPath file, so the rename atomically has all properties set. Copying the metadata from a previous version as in lines 334-350 is brittle ... and might not even work when versioning is somehow disabled. Not necessarily a task for this PR. Please open an issue if you want to tackle this separately.

Reply:
Did it quickly. Should work imo